sapicore/engine/ensemble/__init__.py

```diff
@@ -0,0 +1,100 @@
+""" Ensembles (layers) manage neuron collections by fixing their size, extending select attribute tensors,
+and handling layer-aware operations such as heterogeneous parameter initialization and normalization.
+
+Note
+----
+Ensemble instance attributes are 1D tensors corresponding to potentially heterogeneous parameters
+and states of neurons (tensor elements) whose behavior is prescribed by their parent class
+(e.g., :class:`~engine.neuron.spiking.LIF.LIFNeuron`).
+
+Ensembles can be initialized programmatically or from a configuration YAML containing neuron parameters and their
+generation method (fixed, zipped, grid, or drawn from a distribution). Forward calls invoke the parent neuron class
+method.
+
+Example
+-------
+Initialize an ensemble of basic analog neurons, inheriting default parameters and behavior from
+:class:`~engine.neuron.analog.AnalogNeuron`:
+
+    >>> from sapicore.engine.ensemble.spiking import LIFEnsemble
+    >>> layer = LIFEnsemble(num_units = 10)
+
+See Also
+--------
+:class:`~utils.sweep.Sweep`
+    For built-in parameter sweep utility methods.
+
+"""
+import torch
+from torch import Tensor
+
+from sapicore.engine.neuron import Neuron
+
+
+class Ensemble(Neuron):
+    """Ensemble base class. Provides generic implementations of tensor expansion and parameter diversification.
+
+    Attributes
+    ----------
+    _extensible_props_: tuple[str]
+        The subset of this object's configurable property names (`_config_props_`) that needs to be duplicated
```

Comment:
I know you said that intentionally made it default to `_config_props` and `loggable_props` is always duplicated. But I don't quite follow the reasoning.

Conceptually, `_extensible_props_` is tied to neither, I think!? I understand, as a shortcut, to not have to list them again, that most of the time config properties will be duplicateable, but not always, so `_extensible_props_` default to config props but can be overwritten. And that loggable properties pretty much always will be duplicatable. But I think that's purely by chance and not because `_extensible_props_` is tied to either!?

Hence doesn't it make sense that `_extensible_props_` defaults to the set of `loggable_props` and `_config_props` when None, and otherwise only those listed in it would be considered safe for duplication? I.e. why is `loggable_props` always duplicated?

IMO you should not let it default to either but you'd need to list all of them always. Because if not, like it currently is, you run into this issue: Say you have subclass A set `_extensible_props_` to a specific list meaning only those properties are safe, and subsubclass B set it to None. Then, `if self._extensible_props_ is None` of the subsubclass will return True and it'll return `self._config_props` which includes everything of all its superclasses, which its parent A explicitly listed as unsafe by specifying its `_extensible_props_` list.

In either case, right now you only get the current class `_extensible_props_`, but really you should be iterating the full superclass heirarchy like loggable and configurable does. But, once you do that, you'll have to decide what to do when the hierarchy conflicts with each other, hence why I think in general you should be explicitly listing the safe properties in each class, and not relying on either `_config_props` or `loggable_props`.

Reply:
This ties into my thoughts on having some parent property class that will encapsulate the "iterate over the hierarchy" functionality for any X-able you can conceive of. Plus, the only material difference between loggable and configurable is that the latter is specified once (before the simulation) while the former updates periodically. A job for a future PR, in my opinion.

You won me over re extending _extensible to both configurable and loggable properties. In practice, if you find yourself writing a component class where a non-duplicable property is loggable, it must be some global state that's a function of the entire layer. If you need it for subsequent analysis, you'd compute it post hoc. If you need it to make decisions mid-experiment, it should be part of the logic. Maybe we'll come up with scenarios where extending that flexibility would be warranted, but I'm wary of erring on the side of just-in-case flexibility.

Reply:
Deferring modification of _extensible to a future patch. Low priority compared to other tasks.